src/formState.ts

```diff
@@ -6,28 +6,63 @@ import { assertNever, fail } from "src/utils";
 
 /**
  * Creates a formState instance for editing in a form.
- *
- * @param config
- * @param initValue the initial value from GraphQL, i.e. the Author that we're editing
- * @param initFn a function to adapt the Author "output" object to the form's object.
- * @param opts a hook to add cross-field rules that can't be declared in `config`
  */
-export function useFormState<T, O>(
-  config: ObjectConfig<T>,
-  initValue: O,
-  initFn: (initValue: O) => T,
-  opts?: {
-    addRules?: (state: ObjectState<T>) => void;
-    readOnly?: boolean;
-    /** Fired when the form should auto-save, i.e. after a) blur + b) all fields are valid. */
-    autoSave?: (state: ObjectState<T>) => void;
-  },
-): ObjectState<T> {
-  const { addRules, readOnly = false, autoSave } = opts || {};
+export function useFormState<T, O>(opts: {
+  /** The form configuration, should be a module-level const or useMemo'd. */
+  config: ObjectConfig<T>;
+
+  /**
+   * Provides the form's initial value, as adapted from the `initInput` (if provided).
+   *
+   * If you've not passed the `initInput` key, this function will still be called, just once,
+   * to provide the initial value (i.e. you don't need to useMemo/useCallback the function).
+   *
+   * If you have passed the `initInput` key, this function will be called with the de-optional'd
+   * `initInput` (i.e. so that you don't have to do `undefined` checks while waiting for a GQL
+   * query to load), and then re-called anytime the identity of `initInput` changes.
+   */
+  initFn: (initInput: O) => T;
+
+  /**
+   * The initFn's input value, i.e. from GraphQL the Author that we're editing, so that you can
+   * adapt it to the form's value. Should be stable/useMemo'd.
+   *
+   * If not provided, then `initFn` will always be called.
+   * If provided, then `initFn` will only be called when this is defined.
+   */
+  initInput?: O | null | undefined;
+
+  /** The initial value to use if you pass `initInput`, but it's `undefined`. Defaults to `{}`. */
+  initValueIfUndefined?: T;
```

Comment:
This just feels kind of funky. I'm trying to make guesses as to where this use case would be... maybe if the form was previously loaded and then we navigate back to the page, and we use the cached data as the default value? Though, in that case I think we'd just use the `initFn` to pass the same `query.data` object, and once it's defined then `initFn` is recalled as the identity would've changed, right?
Any use case examples? Maybe for an empty form that we want to provide defaults for? Though, I guess that'd still use the `initFn` too. 🤔

Reply:
I don't disagree, Kolton and I chatted about it for awhile. Search the BP PR for initValueIfUndefined and you'll see where we use it. Probably 4-5 places.

Reply:
Also, the only reason that `initValueIfUndefined` exists is because I wanted to make `initFn` _only called if `inputValue` is defined_. Which is a big change from today. The reason I wanted to change it is code like this:

https://github.com/homebound-team/procurement-frontend/blob/main/src/routes/RequestOverviewTab.tsx#L39

Which is doing a lot of `data?` conditional accesses that could go away.

Similarly, in BP we had a ~4-5 of `initFn: (data) => data ? ...default value... : ({ ...actually use data ... })` that are now pulled apart.

In a way that I _think_ is better...